tproxy: E2E tests

tgross opened the pull request:
_WIP_

gulducat commented on e2e/connect/connect_test.go:
```diff
@@ -120,6 +121,30 @@ func testConnectMultiService(t *testing.T) {
 
+// testConnectTransparentProxy tests the Connect Transparent Proxy integration
+func testConnectTransparentProxy(t *testing.T) {
+	_, cleanup := jobs3.Submit(t, "./input/tproxy.nomad.hcl", jobs3.Timeout(time.Second*60))
+	t.Cleanup(cleanup)
```
lil tidbit: `jobs3`s cleanup automatically, unless instructed not to (d34788896f8892377a9039b81a65abd7a913b3cc)

gulducat commented on e2e/terraform/packer/ubuntu-jammy-amd64/setup.sh:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+ #!/usr/bin/env bash
```
sneaky lil space sneaked in here

gulducat commented on e2e/terraform/packer/ubuntu-jammy-amd64/setup.sh:
```diff
@@ -63,6 +64,25 @@ sudo apt-get install -y \
+# TODO(tgross: replace with downloading the binary from releases.hashicorp.com
+# once the official 1.4.2 release has shipped
+echo "Installing consul-cni plugin"
+sudo apt-get install -y build-essential git curl
```
curl's already installed up on line 30

gulducat commented on e2e/terraform/packer/ubuntu-jammy-amd64/setup.sh:
```diff
@@ -63,6 +64,25 @@ sudo apt-get install -y \
+sudo apt-get install -y build-essential git curl
+
+pushd /tmp
+curl -LO https://go.dev/dl/go1.22.2.linux-amd64.tar.gz
+sudo tar -C /usr/local -xzf go1.22.2.linux-amd64.tar.gz
```
we really didn't already have Go installed on these boxes?  I'd put this under its own little heading, even if it's only currently used for "consul-k8s".

unless you're piling these things together to be also removed in bulk after 1.4.2 is shipped?

tgross replied:
No, we don't generally need toolchains on these machines. And yeah, I was putting them all together so they're easy to find and remove once 1.4.2 ships.

gulducat commented on e2e/connect/connect_test.go:
```diff
@@ -120,6 +121,30 @@ func testConnectMultiService(t *testing.T) {
+	})
+
+	assertServiceOk(t, cc, "count-api-sidecar-proxy")
+	assertServiceOk(t, cc, "count-dashboard-sidecar-proxy")
```
are service healthchecks adequate for confirming all the things we care about with this system?

 * services are up individually
 * one can talk to the other
 * they can't also bypass the mesh?
 * ... other?

tgross replied:
That's a good point... I should check on that dashboard health check to verify that it checks reachability. Bypass is a little harder to verify, because these applications are anemic container images. If reachability works that implies we've called `consul-cni` correctly, because it includes the DNS config. Most of the actual iptables rules details are owned by the plugin. I can see if there's improvements to be made there though.

tgross replied:
Well unfortunately the dashboard health doesn't include that, and the dashboard contents themselves are rendered via Javascript. So we can't even hit the dashboard alloc logs to find the "Fetched count" values:

```
$ nomad alloc logs -task dashboard 5a30
Starting server on http://0.0.0.0:9002
(Pass as PORT environment variable)
Using counting service at http://127.0.0.1:8080
(Pass as COUNTING_SERVICE_URL environment variable)
Starting websocket server...
New client connected
Fetched count 1
Fetched count 2
Fetched count 3
Fetched count 4
```

Because there's nothing to drive the Javascript! If we're going to continue to use countdash as our test application, adding an upstream reachability health check seems to be the way to go. Maybe there's some better choice we could use here, where the front server is proxying directly. I'll look into that.

gulducat replied:
how about `exec`ing from within the task(s)? to exercise network from dashboard to api, from either to world, etc.

tgross replied:
I was going to say the container images are very bare-bones so there's nothing to exercise the network with, but then I remembered they have `busybox` and that has `wget`. Lemme give that a try.

tgross replied:
Assertion for upstream reachability added

shoenig commented on e2e/terraform/packer/ubuntu-jammy-amd64/setup.sh:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+ #!/usr/bin/env bash
```
space?

tgross replied:
:facepalm:

gulducat reviewed (approved):
it's all coming together!  few little nitpicks, and one question about the test assertions.

shoenig commented on e2e/terraform/packer/ubuntu-jammy-amd64/setup.sh:
```diff
@@ -63,6 +64,25 @@ sudo apt-get install -y \
+pushd /tmp
+curl -LO https://go.dev/dl/go1.22.2.linux-amd64.tar.gz
+sudo tar -C /usr/local -xzf go1.22.2.linux-amd64.tar.gz
+git clone https://github.com/hashicorp/consul-k8s.git
```
depth=1?

tgross replied:
Good call.

shoenig reviewed (approved):
LGTM

tgross commented:
After addressing code review comments:

```
$ go test -v -count=1 ./connect
=== RUN   TestConnect_LegacyACLs
=== RUN   TestConnect_LegacyACLs/ConnectDemo
=== RUN   TestConnect_LegacyACLs/ConnectDemoNamespaced
=== RUN   TestConnect_LegacyACLs/ConnectNativeDemo
=== RUN   TestConnect_LegacyACLs/ConnectIngressGatewayDemo
=== RUN   TestConnect_LegacyACLs/ConnectTerminatingGateway
--- PASS: TestConnect_LegacyACLs (93.03s)
    --- PASS: TestConnect_LegacyACLs/ConnectDemo (24.84s)
    --- PASS: TestConnect_LegacyACLs/ConnectDemoNamespaced (25.66s)
    --- PASS: TestConnect_LegacyACLs/ConnectNativeDemo (14.41s)
    --- PASS: TestConnect_LegacyACLs/ConnectIngressGatewayDemo (13.39s)
    --- PASS: TestConnect_LegacyACLs/ConnectTerminatingGateway (14.48s)
=== RUN   Test_serviceOfSIToken
--- PASS: Test_serviceOfSIToken (0.00s)
=== RUN   TestConnect_ClientRestart
    client_test.go:17: skipping test that does nomad agent restart
--- SKIP: TestConnect_ClientRestart (0.00s)
=== RUN   TestConnect
=== RUN   TestConnect/ConnectDemo
=== RUN   TestConnect/ConnectCustomSidecarExposed
=== RUN   TestConnect/ConnectNativeDemo
=== RUN   TestConnect/ConnectIngressGatewayDemo
=== RUN   TestConnect/ConnectMultiIngress
=== RUN   TestConnect/ConnectTerminatingGateway
=== RUN   TestConnect/ConnectMultiService
=== RUN   TestConnect/ConnectTransparentProxy
--- PASS: TestConnect (145.09s)
    --- PASS: TestConnect/ConnectDemo (24.78s)
    --- PASS: TestConnect/ConnectCustomSidecarExposed (16.35s)
    --- PASS: TestConnect/ConnectNativeDemo (14.32s)
    --- PASS: TestConnect/ConnectIngressGatewayDemo (13.29s)
    --- PASS: TestConnect/ConnectMultiIngress (14.36s)
    --- PASS: TestConnect/ConnectTerminatingGateway (14.41s)
    --- PASS: TestConnect/ConnectMultiService (15.43s)
    --- PASS: TestConnect/ConnectTransparentProxy (31.90s)
PASS
ok      github.com/hashicorp/nomad/e2e/connect  238.124s
```
